### PR TITLE
[query/ptypes] remove requiredeness from LegacyEncodedTypeParser and AnnotationImpex

### PR DESCRIPTION
**Pull request by akotlar**

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/AnnotationImpex.scala [hunk 1]
```diff
@@ -25,17 +26,17 @@ object SparkAnnotationImpex {
...
-  def importType(t: DataType): Type = t match {
-    case BooleanType => TBoolean()
-    case IntegerType => TInt32()
-    case LongType => TInt64()
-    case FloatType => TFloat32()
-    case DoubleType => TFloat64()
-    case StringType => TString()
-    case BinaryType => TBinary()
-    case ArrayType(elementType, containsNull) => TArray(importType(elementType).setRequired(!containsNull))
+  def importType(t: DataType): PType = t match {
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/AnnotationImpex.scala]: Want to check: are there any stricter requiredeness rules we can apply here? One thought: if ArrayType.containsNull is false, one interpretation. is that the array itself cannot be null (the other is that the outer array can be missing, but when present contains no nulls).

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/AnnotationImpex.scala]: nope, this is as good a we can do.

File: hail/src/main/scala/is/hail/expr/ir/TableValue.scala [hunk 2]
```diff
@@ -25,8 +25,11 @@ object TableValue {
       RVD.coerce(RVDType(rowType, key), rdd, ctx))
   }
 
-  def apply(ctx: ExecuteContext, rowType:  TStruct, key: IndexedSeq[String], rdd: RDD[Row]): TableValue = {
-    val canonicalRowType = PStruct.canonical(rowType)
+  def apply(ctx: ExecuteContext, rowType:  TStruct, key: IndexedSeq[String], rdd: RDD[Row], rowPType: Option[PStruct] = None): TableValue = {
+    val canonicalRowType = rowPType match {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/TableValue.scala]: more idiomatic -- `rowPType.getOrElse(PStruct.canonical(rowType))`

File: hail/src/main/scala/is/hail/expr/AnnotationImpex.scala (hunk 1 above)
Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/AnnotationImpex.scala]: What are the semantics of containsNull?

Documentation opposite of detailed: https://spark.apache.org/docs/2.1.1/api/java/org/apache/spark/sql/types/ArrayType.html#containsNull()

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/AnnotationImpex.scala]: containsNull means that the elements can be null. So !containsNull means that the element should be required.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/AnnotationImpex.scala]: that's what we do here

Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/AnnotationImpex.scala]: Yeah, I was curious about whether this also gave information about the Array itself (the container). 

So now I understand: the type can always be missing, but children of a type can be  never-null

**Comment by akotlar** [comment]
Should be good

**Pull request merged**